core/arch/instr.h

```diff
@@ -427,6 +427,45 @@ struct _instr_t {
 };     /* instr_t */
 #endif /* DR_FAST_IR */
 
+/**
+ * This note is used to hold emulation specific data which can be set by an
+ * emulation client and read by an instrumentation client. An example of its use
+ * is provided in api/samples/inscount.cpp
+ */
+enum note_type_id {
+    EMULATION_STATE = 1, /* Instruction used to emulate, contains emulation data */
+    EMULATION_HELPER = 2 /* Instruction used to emulate */
+};
+
+/**
+ * TODO: Full doxygen comments will be done as part of upstream pull request,
+ *       when code has been reviewed.
+ */
+typedef struct _note_t {
+    enum note_type_id type_id;
+    struct _note_t *next;
+} note_t;
+
+/**
+ * TODO: Full doxygen comments will be done as part of upstream pull request,
+ *       when code has been reviewed.
+ */
+typedef struct _emulation_state_t {
+    note_t header;
+    app_pc pc;
+    uint raw_instr_bits;
```

Comment:
It sounds like your plan is to keep the original instruction there.  I would vote for instead removing it: i.e., having the app2app changes for emulation produce what will be executed on the hardware.  The emulation client would remove the original app instruction and replace it with the sequence of new instructions that emulate the original, surrounding those with label markers, with the start label pointing to the original (which would be freed in the instru2instru phase or the end of the insertion phase).

Reply:
The logic here is that other types of tools will want to operate on the emulation sequence and not the original.  E.g., a Dr. Memory type tool.  Having the original instruction sitting there yet be removed later seems like it's going to confuse some tools and lead to trouble: it seems better to have the app instr sequence be the final one, and have the original in the metadata.

Reply:
It's not clear form the patch, but the emulation client does remove the original instruction, after its PC and encoding have been saved in emulation_state_t.

One thing that does concern me is what DynamoRIO is free to do with the emulation code between the labels. Will the code be re-arranged/optimised as a result of being placed between two meta instructions (the labels)?

Reply:
Base DR does not perform any re-arranging and only performs "mangling" on branches and some other instructions (for reachability, stolen register, etc.).  This is purely local.  Nothing will be moved across the labels.